server/resp/src/main/java/org/infinispan/server/resp/RespRequestHandler.java

```diff
@@ -1,16 +1,109 @@
 package org.infinispan.server.resp;
 
 import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import org.infinispan.util.concurrent.CompletionStages;
 
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufAllocator;
 import io.netty.buffer.ByteBufUtil;
 import io.netty.channel.ChannelHandlerContext;
 
-public interface RespRequestHandler {
-   default RespRequestHandler handleRequest(ChannelHandlerContext ctx, String type, List<byte[]> arguments) {
+public abstract class RespRequestHandler {
+   protected final CompletionStage<RespRequestHandler> myStage = CompletableFuture.completedStage(this);
+   protected final Function<Object, RespRequestHandler> myFunction = __ -> this;
+   protected final BiFunction<Object, Throwable, RespRequestHandler> myHandler = (__, t) -> this;
+
+   /**
+    * Handles the RESP request returning a stage that when complete notifies the command has completed as well as
+    * providing the request handler for subsequent commands.
+    *
+    * @param ctx Netty context pipeline for this request
+    * @param type The command type
+    * @param arguments The remaining arguments to the command
+    * @return stage that when complete returns the new handler to instate. This stage <b>must</b> be completed on the event loop
```

Comment:
The returned stage having to be completed on the event loop is very important to all of this working properly.